File: BE/controllers/subscription.controller.js
const db = require('../database/db');

// Get all plans
exports.getPlans = async (req, res) => {
  try {
    const [plans] = await db.query('SELECT * FROM subscription_plans ORDER BY id');
    res.json(plans);
  } catch (err) {
    console.error('Get plans error:', err);
    res.status(500).json({ message: 'Server error' });
  }
};

// Get current subscription
exports.getCurrentPlan = async (req, res) => {
  try {
    const [company] = await db.query(
      'SELECT subscription_plan, task_limit, tasks_used, subscription_expiry FROM companies WHERE id = ?',
      [req.user.id]
    );

    if (company.length === 0) {
      return res.status(404).json({ message: 'Company not found' });
    }

    const [planDetails] = await db.query(
      'SELECT * FROM subscription_plans WHERE plan_name = ?',
      [company[0].subscription_plan]
    );

    res.json({
      ...company[0],
      plan_details: planDetails[0] || null
    });
  } catch (err) {
    console.error('Get current plan error:', err);
    res.status(500).json({ message: 'Server error' });
  }
};

// Upgrade plan
exports.upgradePlan = async (req, res) => {
  try {
    const { plan } = req.body;

    if (!['basic', 'standard', 'premium'].includes(plan)) {
      return res.status(400).json({ message: 'Invalid plan' });
    }

    const [planDetails] = await db.query('SELECT * FROM subscription_plans WHERE plan_name = ?', [plan]);
    if (planDetails.length === 0) {
      return res.status(400).json({ message: 'Plan not found' });
    }

    // Set expiry to 30 days from now
    const expiry = new Date();
    expiry.setDate(expiry.getDate() + 30);

    await db.query(
      `UPDATE companies SET subscription_plan = ?, task_limit = ?, subscription_expiry = ?
       WHERE id = ?`,
      [plan, planDetails[0].task_limit, expiry, req.user.id]
    );

    res.json({ message: `Plan upgraded to ${plan} successfully` });
  } catch (err) {
    console.error('Upgrade plan error:', err);
    res.status(500).json({ message: 'Server error' });
  }
};

File: BE-Python/controllers/subscription_controller.py
from datetime import datetime, timedelta
from flask import request, jsonify, g
from database.db import query, query_one, execute

def get_plans():
    try:
        plans = query('SELECT * FROM subscription_plans ORDER BY id')
        return jsonify(plans)
    except Exception as e:
        print(f'Get plans error: {e}')
        return jsonify({'message': 'Server error'}), 500

def get_current_plan():
    try:
        company = query_one(
            'SELECT subscription_plan, task_limit, tasks_used, subscription_expiry FROM companies WHERE id = %s',
            (g.user['id'],)
        )

        if not company:
            return jsonify({'message': 'Company not found'}), 404

        plan_details = query_one(
            'SELECT * FROM subscription_plans WHERE plan_name = %s',
            (company['subscription_plan'],)
        )

        result = {**company, 'plan_details': plan_details}
        return jsonify(result)

    except Exception as e:
        print(f'Get current plan error: {e}')
        return jsonify({'message': 'Server error'}), 500

def upgrade_plan():
    try:
        data = request.get_json()
        plan = data.get('plan')

        if plan not in ('basic', 'standard', 'premium'):
            return jsonify({'message': 'Invalid plan'}), 400

        plan_details = query_one('SELECT * FROM subscription_plans WHERE plan_name = %s', (plan,))
        if not plan_details:
            return jsonify({'message': 'Plan not found'}), 400

        company = query_one('SELECT subscription_plan FROM companies WHERE id = %s', (g.user['id'],))
        if not company:
            return jsonify({'message': 'Company not found'}), 404

        if company['subscription_plan'] == plan:
            return jsonify({'message': 'You are already on this plan'}), 400

        # Check if there is already a pending request for this plan
        existing = query_one(
            "SELECT id FROM upgrade_requests WHERE company_id = %s AND requested_plan = %s AND status = 'pending'",
            (g.user['id'], plan)
        )
        if existing:
            return jsonify({'message': 'You already have a pending request for this plan. Please wait for admin approval.'}), 400

        # Create upgrade request instead of direct upgrade
        execute(
            'INSERT INTO upgrade_requests (company_id, current_plan, requested_plan) VALUES (%s, %s, %s)',
            (g.user['id'], company['subscription_plan'], plan)
        )

        return jsonify({'message': 'Upgrade request submitted successfully. Please wait for admin approval.'})

    except Exception as e:
        print(f'Upgrade plan error: {e}')
        return jsonify({'message': 'Server error'}), 500
